### PR TITLE
Protected Sheet and WorkBook

### PR DESCRIPTION
**Pull request by DevOlegKosarev**
+ Protected Sheet
+ Protected WorkBook.

Removing special characters known to me from the sheet name. Since if you do not delete these characters, then Excel says that the sheet is broken and offers to restore it.
___________
At the moment I'm working on password protection for workbook and sheet. But so far without success, since it is not yet clear to me how the excel password is generated.

### CODE REVIEW

File: src/SimpleXLSXGen.php [hunk 1]
```diff
@@ -40,6 +40,8 @@ class SimpleXLSXGen
     protected $keywords;
     protected $category;
     protected $lastModifiedBy;
+    protected bool $prtoectedWorkbook;
```
Comment (shuchkin) [inline src/SimpleXLSXGen.php:43]: remove type decl, we use PHP 7.0 (prop types added in 7.4)
rename to $isProtectedWorkbook;

File: src/SimpleXLSXGen.php [hunk 2]
```diff
@@ -40,6 +40,8 @@ class SimpleXLSXGen
...
+    protected array $prtoectedSheet;
```
Comment (shuchkin) [inline src/SimpleXLSXGen.php:44]: move flags to ```$sheets```,  ```$this->sheets[$idx']['protect']```

File: src/SimpleXLSXGen.php [hunk 3]
```diff
@@ -1078,9 +1097,15 @@ public function setLastModifiedBy($lastModifiedBy)
         return $this;
     }
 
-    public function setLastModifiedBy($lastModifiedBy)
+    public function setProtectedWorkBook()
```
Comment (shuchkin) [inline src/SimpleXLSXGen.php:1100]: ?

File: src/SimpleXLSXGen.php [hunk 4]
```diff
@@ -935,6 +945,15 @@ protected function _sheetToXML($idx, $template)
...
 
+        $protectedSheet = '';
+        if (is_array($this->sheets[$idx]['isPrtoectedSheet']) and count($this->sheets[$idx]['isPrtoectedSheet']) > 0) {
+            $premissions = 'sheet="1" ';
+            foreach ($this->sheets[$idx]['isPrtoectedSheet'] as $namePremissions => $valuePremissions) {
+                $premissions .= $namePremissions . '="' . (int)$valuePremissions . '" ';
+            }
+            $protectedSheet = '<sheetProtection ' . $premissions . '/>';
+        }
+
```
[the remark this answers is not in the document]
Comment (DevOlegKosarev) [reply src/SimpleXLSXGen.php:956]: me is commit https://github.com/DevOlegKosarev/simplexlsxgen/commit/e4eddd6d1cebbd219436d8b543176939a6a617a2 this and write...

File: src/SimpleXLSXGen.php [hunk 5]
```diff
@@ -340,12 +346,16 @@ protected function _write($fh)
...
+                    $prtoectedWorkbook = '';
+                }
                 $s = '';
                 foreach ($this->sheets as $k => $v) {
                     $s .= '<sheet name="' . $this->esc($v['name']) . '" sheetId="' . ($k + 1) . '" r:id="rId' . ($k + 1) . '"/>';
                 }
-                $search = ['{SHEETS}', '{APP}'];
-                $replace = [$s, $this->esc($this->application)];
+                $search = ['{SHEETS}', '{APP}', '{PROTECTED_WORKBOOK}'];
+                $replace = [$s, $this->esc($this->application), $prtoectedWorkbook];
```
[the remark this answers is not in the document]
Comment (DevOlegKosarev) [reply src/SimpleXLSXGen.php:358]: yes me is fix this

File: src/SimpleXLSXGen.php [hunk 6]
```diff
@@ -340,12 +346,16 @@ protected function _write($fh)
                 $this->_writeEntry($fh, $cdrec, $cfilename, $template);
                 $entries++;
             } elseif ($cfilename === 'xl/workbook.xml') {
+                $prtoectedWorkbook = '<workbookProtection lockStructure="1"/>';
+                if ($this->isProtectedWorkbook == false) {
+                    $prtoectedWorkbook = '';
+                }
```
[the remark this answers is not in the document]
Comment (DevOlegKosarev) [reply src/SimpleXLSXGen.php:352]: I think it's not important how to write this is a matter of taste.

File: src/SimpleXLSXGen.php [hunk 7]
```diff
@@ -102,6 +104,9 @@ public function __construct()
         $this->lastModifiedBy = 'Sergey Shuchkin <sergey.shuchkin@gmail.com>';
         $this->application = __CLASS__;
 
+        $this->isProtectedWorkbook = false;
+        $this->isPrtoectedSheet = [];
```
[the remark this answers is not in the document]
Comment (DevOlegKosarev) [reply src/SimpleXLSXGen.php:108]: I will check this moment. But if I have an error there, it is very strange that it works correctly.

**Comment by DevOlegKosarev** [comment]
All suggestions and comments I will check and do but a little later and send in the final comite. I need about 1-2 days. Now I am very busy with my main job.

**Comment by shuchkin** [comment]
Oleg, the code is crude now, maybe because of the rush

* Функция нужная, но код пока грязный, я обычно смотрю как это можно сделать на VBA

**Pull request closed without merging**